llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -4585,33 +4586,43 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *bv, Function *f, Bas
   }
 
   case OpVectorShuffle: {
-    // NOTE: LLVM backend compiler does not well handle "shufflevector"
-    // instruction. So we avoid generating "shufflevector" and use the
-    // combination of "extractelement" and "insertelement" as a substitute.
     auto vs = static_cast<SPIRVVectorShuffle *>(bv);
 
     auto v1 = transValue(vs->getVector1(), f, bb);
     auto v2 = transValue(vs->getVector2(), f, bb);
 
-    auto vec1CompCount = vs->getVector1ComponentCount();
     auto newVecCompCount = vs->getComponents().size();
+    auto vec1CompCount = vs->getVector1ComponentCount();
+    auto vec2CompCount = vs->getVector2ComponentCount();
 
-    IntegerType *int32Ty = IntegerType::get(*m_context, 32);
-    Type *newVecTy = FixedVectorType::get(cast<VectorType>(v1->getType())->getElementType(), newVecCompCount);
-    Value *newVec = UndefValue::get(newVecTy);
+    llvm::SmallVector<int, 4> mask(vs->getComponents().begin(), vs->getComponents().end());
 
-    for (size_t i = 0; i < newVecCompCount; ++i) {
-      auto comp = vs->getComponents()[i];
-      if (comp < vec1CompCount) {
-        auto newVecComp = ExtractElementInst::Create(v1, ConstantInt::get(int32Ty, comp), "", bb);
-        newVec = InsertElementInst::Create(newVec, newVecComp, ConstantInt::get(int32Ty, i), "", bb);
-      } else {
-        auto newVecComp = ExtractElementInst::Create(v2, ConstantInt::get(int32Ty, comp - vec1CompCount), "", bb);
+    // If vectors are not the same size then extend smaller one
+    // to be the same as bigger. This is required by llvm::shufflevector.
+    if (v1->getType() != v2->getType()) {
+
+      auto smallerVec = std::min(vec1CompCount, vec2CompCount);
+      auto biggerVec = std::max(vec1CompCount, vec2CompCount);
+
+      // Create temporary mask used to extend size of smaller vector.
+      llvm::SmallVector<int> extendedVecMask = llvm::createSequentialMask(0, smallerVec, (biggerVec - smallerVec));
```

Comment:
Just a nit. Remove "llvm::" prefix in .cpp file.

Reply:
done